### PR TITLE
fix(storage/mergeQuerier): fix a data race

### PR DESCRIPTION
**Pull request by machine424**
fixes https://github.com/prometheus/prometheus/issues/14723

### CODE REVIEW

File: storage/merge.go [hunk 1]
```diff
@@ -153,13 +153,21 @@ func (q *mergeGenericQuerier) Select(ctx context.Context, sortSeries bool, hints
 	)
 	// Schedule all Selects for all queriers we know about.
 	for _, querier := range q.queriers {
+		// copy the matchers as some queriers may alter the slice.
+		// See https://github.com/prometheus/prometheus/issues/14723
+		//matchersCopy := make([]*labels.Matcher, len(matchers))
+		//copy(matchersCopy, matchers)
```
Comment (machine424) [inline storage/merge.go:159]: I think there is no need to deep copy the matchers themselves?

File: storage/merge.go [hunk 2]
```diff
@@ -153,13 +153,21 @@ func (q *mergeGenericQuerier) Select(ctx context.Context, sortSeries bool, hints
 	)
 	// Schedule all Selects for all queriers we know about.
 	for _, querier := range q.queriers {
+		// copy the matchers as some queriers may alter the slice.
```
Comment (machine424) [inline storage/merge.go:156]: I don't think that preventing blockQuerier from altering the matchers or assuming no other querier will respect that in the future is a good/safe solution.

File: tsdb/querier_test.go [hunk 3]
```diff
@@ -3787,3 +3787,29 @@ func (m mockReaderOfLabels) Series(storage.SeriesRef, *labels.ScratchBuilder, *[
...
+		labels.MustNewMatcher(labels.MatchRegexp, "baz", ".*"),
+		labels.MustNewMatcher(labels.MatchEqual, "foo", "bar"),
+	}
+	matchers := append([]*labels.Matcher{}, originalMatchers...)
+
+	mergedQuerier := storage.NewMergeQuerier([]storage.Querier{p}, []storage.Querier{s}, storage.ChainedSeriesMerge)
+	defer mergedQuerier.Close()
+	mergedQuerier.Select(context.Background(), false, nil, matchers...)
+
+	require.Equal(t, originalMatchers, matchers)
```
Comment (machine424) [inline tsdb/querier_test.go:3814]: needed as the race detector is disabled on tests currently.

**Comment by machine424** [comment]
test is failing https://github.com/prometheus/prometheus/actions/runs/11038907137/job/30663187884?pr=14983#step:6:656, I'll push the fix now.

File: storage/merge.go [hunk 4]
```diff
@@ -153,13 +153,18 @@ func (q *mergeGenericQuerier) Select(ctx context.Context, sortSeries bool, hints
...
+		// See https://github.com/prometheus/prometheus/issues/14723
+		matchersCopy := make([]*labels.Matcher, len(matchers))
```
[the remark this answers is not in the document]
Comment (machine424) [reply storage/merge.go:158]: Yep, I left a comment about that here https://github.com/prometheus/prometheus/pull/14983#discussion_r1775736242, I'm concerned that approach isn't future proof (no guarantee that we'll not alter that slice somewhere else), however, if you believe it’s worth the risk, we can copy the slice before sorting n `blockQuerier`.

copying is ~8bytes per matcher (+some constant bytes), given the common use cases I'm aware of (theyre are no benchmarks that emphasize 10,000 matchers use cases e.g.) for the read API, I think that overhead could be negligeable; for prometheus, supposing `queriers#=2`, `queryConcurrency=20`, `avg_select_duration=100ms` and 5 matchers per query, we'll be producing ~`2*20*5*8=1600bytes` of garbage every `100ms` = ~`16kB/s` in the worst case.

We could also consider pooling to manage this, but avoiding additional complexity would be preferable.

[the remark this answers is not in the document]
Comment (machine424) [reply storage/merge.go:158]: cc @bwplotka in case this went unnoticed.

**Review by bboreham — APPROVED** [review]
Code looks ok to me.  I fired up PromBench to see if the allocation shows up.
(But the base is a month off so may not be accurate)

File: storage/merge.go [hunk 5]
```diff
@@ -153,13 +153,18 @@ func (q *mergeGenericQuerier) Select(ctx context.Context, sortSeries bool, hints
...
+		copy(matchersCopy, matchers)
```
[the remark this answers is not in the document]
Comment (machine424) [reply storage/merge.go:159]: for better readability? AFAIK, `copy()` is generally faster than `append()` (which is used by `slices.Clone()`)

**Comment by machine424** [comment]
> Code looks ok to me. I fired up PromBench to see if the allocation shows up.
> (But the base is a month off so may not be accurate)

Thanks, yes we can see a lot of discrepancy, I can rebase and retrigger the bench if you want.

**Comment by machine424** [comment]
Now that I'm thinking about it, AFAIU, prombench doesn't make use of `remote_read`, the only way we can make use of the `mergeGenericQuerier` within Prometneus itself.

**Pull request merged**